File: 111.Minimum-Depth-of-Binary-Tree.py
# https://leetcode.com/problems/minimum-depth-of-binary-tree/
#
# algorithms
# Easy (35.41%)
# Total Accepted:    300,329
# Total Submissions: 848,240
# beats 88.11% of python submissions


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        res = [float('inf')]

        def recursive(node, depth):
            if not node.left and not node.right:
                res[0] = min(res[0], depth + 1)
                return

            if depth >= res[0]:
                return

            if node.left:
                recursive(node.left, depth + 1)
            if node.right:
                recursive(node.right, depth + 1)

        recursive(root, 0)

        return res[0]

File: 111.Minimum-Depth-of-Binary-Tree.java
// https://leetcode.com/problems/minimum-depth-of-binary-tree/
//
// algorithms
// Easy (36.41%)
// Total Accepted: 356,218
// Total Submissions: 978,343
// beats 100.0% of java submissions


/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {

    private static int res;

    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        res = Integer.MAX_VALUE;

        recursive(root, 0);

        return res;
    }

    private void recursive(TreeNode node, int level) {
        if (node.left == null && node.right == null) {
            res = Math.min(res, level + 1);
            return;
        }

        if (node.left != null) {
            recursive(node.left, level + 1);
        }

        if (node.right != null) {
            recursive(node.right, level + 1);
        }
    }
}